feat: Add dequeue strategies support for queue scheduling

kingeasternsun opened the pull request:
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:

This PR introduces dequeue strategies for Volcano queues, allowing administrators to control how jobs are scheduled within queues when resources are limited. 

**Key Features:**
Defined DequeueStrategy type with two supported strategies:

1. **FIFO Strategy**

   * The scheduler repeatedly tries to dequeue the first task in the queue.
   * If the first task cannot be dequeued (e.g., insufficient resources), it will keep retrying without skipping to the next task.

2. **Traversal Strategy**

   * If the first task cannot be dequeued, the scheduler skips it and attempts to dequeue the next jobs in the queue sequentially.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
Notice that do not add `Fixes` if the issue is associated with multiple PRs.
-->
Fixes #4572 #4690 #3950

#### Special notes for your reviewer:

1. **Backward Compatibility**: This change is fully backward compatible. Existing queues without dequeue strategy annotations will continue to use the traverse strategy (current default behavior).

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required.
-->
```release-note
support FIFO dequeue strategies for Volcano queues..
```

---

**Additional Context:**

This feature addresses common scheduling scenarios where different queues require different behavior patterns:

- **Production queues** often need strict ordering and can tolerate some resource underutilization
- **Development/testing queues** typically prioritize maximum resource utilization over strict ordering

The implementation uses a unified scheduling loop with conditional logic, making it maintainable and efficient while providing clear behavioral differences between the two strategies.

kingeasternsun commented on pkg/scheduler/actions/allocate/allocate.go:
```diff
@@ -208,9 +207,15 @@ func (alloc *Action) allocateResources(queues *util.PriorityQueue, jobsMap map[a
 			stmt.Commit()
 		}
 
+		// If the dequeue strategy is FIFO, put the queue back to priority queue only after job's resource allocating successfully
+		// if stmt is nil, it means the job's resource allocating failed, so we should not put the queue back to scheduler the rest of jobs in the queue
```
Fixed.

kingeasternsun commented on pkg/scheduler/actions/allocate/dequeue_strategy_test.go:
```diff
@@ -0,0 +1,394 @@
+}
+
+// Helper function to build a queue with dequeue strategy
+func buildQueueWithStrategy(name string, weight int32, resource *api.Resource, strategy string) *schedulingv1.Queue {
```
Fixed.

kingeasternsun commented on pkg/scheduler/api/queue_info.go:
```diff
@@ -47,11 +64,29 @@
+			switch strategy {
+			case DequeueStrategyFIFO, DequeueStrategyTraverse:
+				dequeueStrategy = strategy
+			default:
+				// Invalid strategy, use default
+				dequeueStrategy = DefaultDequeueStrategy
+			}
```
Fixed.

kingeasternsun commented on pkg/scheduler/actions/allocate/dequeue_strategy_test.go:
```diff
@@ -0,0 +1,394 @@
+	}
+
+	traverseQueue := &api.QueueInfo{
+		UID:             api.QueueID("test-traverse-queue"),
+		Name:            "test-traverse-queue",
+		DequeueStrategy: api.DequeueStrategyTraverse,
+	}
+
+	// Test FIFO behavior: when stmt is nil, should not push queue back
+	t.Run("FIFO strategy with failed job", func(t *testing.T) {
+		if fifoQueue.DequeueStrategy == api.DequeueStrategyFIFO {
+			// This simulates the condition: stmt == nil (job scheduling failed)
+			// In this case, the queue should NOT be pushed back for FIFO strategy
+			t.Log("FIFO strategy correctly blocks queue when job scheduling fails")
+		}
+	})
+
+	// Test Traverse behavior: queue should always be pushed back
+	t.Run("Traverse strategy with failed job", func(t *testing.T) {
+		if traverseQueue.DequeueStrategy == api.DequeueStrategyTraverse {
+			// This simulates the condition: stmt == nil (job scheduling failed)
+			// In this case, the queue should still be pushed back for traverse strategy
+			t.Log("Traverse strategy correctly allows queue to continue when job scheduling fails")
+		}
+	})
+
+	// Test default behavior
+	defaultQueue := &api.QueueInfo{
+		UID:             api.QueueID("test-default-queue"),
+		Name:            "test-default-queue",
+		DequeueStrategy: "", // Empty string should default to traverse
+	}
+
+	t.Run("Default strategy behavior", func(t *testing.T) {
+		// Empty or invalid strategy should behave like traverse
+		if defaultQueue.DequeueStrategy != api.DequeueStrategyFIFO {
+			t.Log("Default strategy correctly behaves like traverse")
+		}
+	})
+}
```
Fixed.

Monokaix commented on pkg/scheduler/actions/allocate/allocate.go:
```diff
@@ -208,6 +207,11 @@ func (alloc *Action) allocateResources(queues *util.PriorityQueue, jobsMap map[a
 
+		// For FIFO, if a job fails to allocate (stmt == nil), block the queue by not re-queuing it.
+		// This prevents subsequent jobs in the same queue from being scheduled in the current cycle.
+		if queue.DequeueStrategy == api.DequeueStrategyFIFO && stmt == nil {
```
Seems if a job failed to allocate, the whole queue will not be re-enqueued, this is not reasonable.

kingeasternsun replied:
Thanks for your comments， I'll fix this.

JesseStutler commented:
good feature, I think it's useful, also needed by https://github.com/volcano-sh/volcano/issues/4690

JesseStutler commented on pkg/scheduler/api/queue_info.go:
```diff
@@ -24,11 +24,29 @@ package api
+const (
+	// DequeueStrategyFIFO means if the first job in queue cannot be dequeued,
+	// the system will repeatedly try to dequeue the first job without skipping
+	DequeueStrategyFIFO = "fifo"
```
I think could we set `dequeueStrategy` as a standard field of the queue and put these consts into apis repo? I think it worth it since many users have reported need it. Also /cc @hwdef @hzxuzhonghu @kevin-wangzefeng @wangyang0616 @lowang-bh

kingeasternsun replied:
Agree.

kingeasternsun replied:
@JesseStutler @hwdef  I have set dequeueStrategy as a standard field of the queue https://github.com/volcano-sh/apis/pull/200 ,  Please review this thanks.

kingeasternsun commented on docs/user-guide/dequeue_strategies.md:
```diff
@@ -0,0 +1,211 @@
+**Behavior**: When a job in the queue fails to schedule due to insufficient resources, the entire queue is blocked for the current scheduling cycle. The scheduler will not attempt to schedule any other jobs from this queue until one of the following occurs:
+- The failed job is removed or modified.
+- A new scheduling cycle begins, which will retry the failed job.
+- Another higher-priority job has been added to the queue.
```
Thanks for the review! I think we can support StrictFIFO, SoftFIFO, and the Traverse strategy simultaneously.

JesseStutler commented:
Please rebase your codes, BTW, 
<img width="1309" height="419" alt="image" src="https://github.com/user-attachments/assets/72eaf89d-ba14-47be-a5e3-3a8c4bcad7c6" />, just a reminder, when writing commits, please clearly describe the changes made, and avoid vague information such as "fix" or "add crd". @kingeasternsun

JesseStutler commented:
@kingeasternsun Could you please resolve the conflicts? And we think we could have a discussion to push forward it again

kingeasternsun commented:
> @kingeasternsun Could you please resolve the conflicts? And we think we could have a discussion to push forward it again

OK.

kingeasternsun committed:
dequeue strategy

Signed-off-by: kingeasternsun <kingeasternsun@gmail.com>

kingeasternsun commented:
> @kingeasternsun Could you please resolve the conflicts? And we think we could have a discussion to push forward it again

Thanks for the review. @Monokaix  @JesseStutler @hwdef 

A lot has changed since the initial submission of this PR, including both production requirements and the Volcano codebase itself. Therefore, I rewrote most of the implementation to align with the current design.

Changes in the latest revision include:

* Added `PriorityFIFO` strategy.
* Recorded failure reasons in both Events and Conditions for better observability.
* Refactored the implementation based on the latest Volcano architecture.

Please review the updated version. Thanks!

kingeasternsun committed:
dequeue strategy lint

Signed-off-by: kingeasternsun <kingeasternsun@gmail.com>

hwdef commented:
Thanks for your contribution. I've been very busy lately, but I'll review this code this week.

hwdef commented on docs/design/dequeue-strategy.md:
```diff
@@ -0,0 +1,116 @@
+
+| Value | Constant | Behavior on allocation failure |
+|-------|----------|--------------------------------|
+| `traverse` | `DequeueStrategyTraverse` | No blocking; other jobs keep trying |
```
What is the origin of `traverse`?. I prefer `besteffort`

hzxuzhonghu commented on pkg/scheduler/actions/allocate/allocate.go:
```diff
@@ -283,6 +315,11 @@ func (alloc *Action) organizeJobWorksheet(job *api.JobInfo) *JobWorksheet {
+	// For FIFO/PriorityFIFO strategies: track the highest failed priority per queue.
+	// PriorityFIFO blocks jobs with priority lower than the threshold.
+	// FIFO blocks jobs with priority lower than or equal to the threshold.
+	failedPriority := map[api.QueueID]*int32{}
```
`failedPriority` exists only during allocate. The default pipeline runs backfill afterward, and backfill independently binds BestEffort tasks without consulting the dequeue strategy, so a lower FIFO job can still run in the same cycle after the head job fails. Preserve blocked queues across actions or enforce the strategy in backfill as well.

hzxuzhonghu reviewed (changes requested):
The FIFO block does not survive the full scheduling cycle.
